File: 0352-data-stream-as-disjoint-intervals/0352-data-stream-as-disjoint-intervals.py
from sortedcontainers import SortedDict

class SummaryRanges:

    def __init__(self):
        self.map = SortedDict()

    def addNum(self, value: int) -> None:
        if value in self.map: return

        right = self.map.bisect_left(value)
        left = right - 1

        keyRight = self.map.peekitem(right)[0] if right < len(self.map) else -1
        keyLeft = self.map.peekitem(left)[0] if left >= 0 else -1
        
        if (keyRight != -1 and keyLeft != -1 and self.map[keyLeft] + 1 == value and value + 1 == keyRight):
            self.map[keyLeft] = self.map[keyRight]
            self.map.pop(keyRight)
        elif (right != -1 and keyRight == value + 1):
            self.map[value] = self.map.pop(keyRight)
        elif (left != -1 and self.map[keyLeft] + 1 >= value):
            self.map[keyLeft] = max(self.map[keyLeft], value)
        else:
            self.map[value] = value

    def getIntervals(self) -> List[List[int]]:
        ans = []        
        for start in self.map:
            ans.append([start, self.map[start]])
        return ans
        

# Your SummaryRanges object will be instantiated and called as such:
# obj = SummaryRanges()
# obj.addNum(value)
# param_2 = obj.getIntervals()

File: 0352-data-stream-as-disjoint-intervals/0352-data-stream-as-disjoint-intervals.cpp
class SummaryRanges {
    set<int> nums;
public:
    SummaryRanges() {
        
    }
    
    void addNum(int value) {
        nums.insert(value);
    }
    
    vector<vector<int>> getIntervals() {
        vector<vector<int>> ans;
        int start = -1, end = -1;
        bool isFirst = true;

        for (const auto &num : nums) {
            if (isFirst) {
                start = end = num;
                isFirst = false;
                continue;
            }

            if (end + 1 == num) {
                end = num;
                continue;
            }

            vector<int> interval;
            interval.push_back(start);
            interval.push_back(end);
            ans.push_back(interval);

            start = end = num;
        }

        if (!isFirst) {
            vector<int> interval;
            interval.push_back(start);
            interval.push_back(end);
            ans.push_back(interval);
        }

        return ans;
    }
};

/**
 * Your SummaryRanges object will be instantiated and called as such:
 * SummaryRanges* obj = new SummaryRanges();
 * obj->addNum(value);
 * vector<vector<int>> param_2 = obj->getIntervals();
 */